### PR TITLE
[SPARK-42884][CONNECT] Add Ammonite REPL integration

### PR DESCRIPTION
**Pull request by hvanhovell**
### What changes were proposed in this pull request?
This PR adds Ammonite REPL integration for Spark Connect. This has a couple of benefits:
- It makes it a lot less cumbersome for users to start a spark connect REPL. You don't have to add custom scripts, and you can use `coursier` to launch a fully function REPL for you.
- It adds REPL integration for to the actual build. This makes it easier to validate the code we add is actually working.

### Why are the changes needed?
A REPL is arguably the first entry point for a lot of users.

### Does this PR introduce _any_ user-facing change?
Yes it adds REPL integration.

### How was this patch tested?
Added tests for the command line parsing. Manually tested the REPL.

### CODE REVIEW

**Review by dongjoon-hyun — DISMISSED** [review]
Hi, @hvanhovell , @cloud-fan , @HyukjinKwon . I have a few questions.
- Are we sure that we provide a different Scala REPL environment in Apache Spark officially?
- If this aims to replace the traditional Scala shell eventually, did we discuss this in the mailing list officially?

To be clear, I'm not against this new improvement. I believe this is worthy of being shared in the mailing list widely.

**Comment by hvanhovell** [comment]
@dongjoon-hyun officially is a bit a broad term. As far as I am concerned ammonite is just a way to use the connect JVM client, it is not meant as a change for all of Spark (although I do think it might make sense). We use other libraries all the time, and I don't think we have ever discussed those in the mailing list.

As it stands there are two reasons for using ammonite:
- It is a better and easier to use REPL than the Scala REPL ever was.
- It is easier to customize. Connect is a good illustration, for the Scala REPL I will probably need to the same thing as we did for the spark REPL project and that is fork the source code. In ammonite this was a minor modification.

I do intent to add support for the regular REPL at some point, this was just easier to get started with.

If you feel strongly about this then I am happy to send a message to the dev-list.

**Comment by dongjoon-hyun** [comment]
Ya, I'm not against this nice improvement. Just shoot one email to the dev mailing list to give a headup. That's what I'm thinking that we need.

**Review by zhenlineo — COMMENTED** [review]
LGTM on the code.

**Comment by hvanhovell** [comment]
@dongjoon-hyun I send the email.

**Comment by dongjoon-hyun** [comment]
Thank you so much, @hvanhovell .

File: connector/connect/client/jvm/src/test/scala/org/apache/spark/sql/connect/client/SparkConnectClientBuilderParseTestSuite.scala [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+  argumentTest("port", "1506", _.port.toString)
+  argumentTest("token", "azbycxdwev1234567890", _.token.get)
+  argumentTest("user_id", "U1238", _.userId.get)
+  argumentTest("user_name", "alice", _.userName.get)
+  argumentTest("user_agent", "MY APP", _.userAgent)
+
+  test("Argument - remote") {
+    val builder =
+      build("--remote", "sc://srv.apache.org/;user_id=x127;user_name=Q;token=nahnah;param1=x")
+    assert(builder.host == "srv.apache.org")
```
Comment (gerashegalov) [inline connector/connect/client/jvm/src/test/scala/org/apache/spark/sql/connect/client/SparkConnectClientBuilderParseTestSuite.scala:53]: here and elsewhere probably should use `===` consistently  like on L35 for more illustrative exception messages?

File: connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/application/ConnectRepl.scala [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+        case NonFatal(e) =>
+          // scalastyle:off println
+          println(s"""
+             |$name
+             |${e.getMessage}
+             |${SparkConnectClientParser.usage()}
+             |""".stripMargin)
+          // scalastyle:on println
+          System.exit(1)
+          return
```
Comment (gerashegalov) [inline connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/application/ConnectRepl.scala:66]:
```suggestion
          sys.exit(1)
```

**Comment by LuciferYang** [comment]
Interesting!

**Review by dongjoon-hyun — APPROVED** [review]

File: connector/connect/bin/spark-connect-scala-client [hunk 3]
```diff
@@ -46,6 +45,4 @@ build/sbt "${SCALA_ARG}" "sql/package;connect-client-jvm/assembly"
 CONNECT_CLASSPATH="$(build/sbt "${SCALA_ARG}" -DcopyDependencies=false "export connect-client-jvm/fullClasspath" | grep jar | tail -n1)"
 SQL_CLASSPATH="$(build/sbt "${SCALA_ARG}" -DcopyDependencies=false "export sql/fullClasspath" | grep jar | tail -n1)"
 
-INIT_SCRIPT="${SPARK_HOME}"/connector/connect/bin/spark-connect-scala-client.sc
-
-exec "${SCALA_BIN}" -cp "$CONNECT_CLASSPATH:$SQL_CLASSPATH" -i $INIT_SCRIPT
+exec java -cp "$CONNECT_CLASSPATH:$SQL_CLASSPATH" org.apache.spark.sql.application.ConnectRepl "$@"
```
Comment (gerashegalov) [inline connector/connect/bin/spark-connect-scala-client:48]: did not notice it earlier, Spark scripts prioritize JAVA_HOME https://github.com/apache/spark/blob/04202b3fdda1ac7aea7d7e5688487ce749ad2f79/bin/spark-class#L27-L36 . Maybe there is a way to leverage this script here to begin with?

**Comment by hvanhovell** [comment]
Merging this one.

File: connector/connect/bin/spark-connect-scala-client (hunk 3 above)
Comment (hvanhovell) [reply connector/connect/bin/spark-connect-scala-client:48]: Sure, let me do that in a follow-up.

**Pull request closed without merging**